Improve mark type sections

binste opened the pull request:
This PR addresses the points raised by @joelostblom in #2607 and #2578 (the later can be closed after this PR). In addition, I went through all mark pages to fix any bugs or vega-lite specific parts I could find. I think the pages are now in pretty good shape but it would certainly be nice if someone else can take a thorough look as well. CC @mattijn in case you want to take a look as well.

I have not yet included the sections with the interactive sliders from the vega-lite documentation. Still need to figure out how to best accomplish this. I'll try the suggestions from comment 4 in #2578 and add it to this PR later on or then in a new one.

You can view the updated documentation at https://binste.github.io/altair-docs/ with the exception of the charts in the geoshape section as I did not have geopandas installed. As a sidenote, where would be a good place to add a note that this is now necessary to build the documentation? Maybe we should add it in `requirements_dev.txt` so it is also included in the docbuild workflow?

binste commented:
Just pushed a new commit which adds the interactive slider chart to the Arc page, see https://binste.github.io/altair-docs/user_guide/marks/arc.html :) Quite excited that it works, I think it's a very handy feature :) 

What do you think of the approach? I pulled the vega-lite specification from their documentation and load it from a separate file with the `raw` directive. I also considered using the `altair.VegaLite` class as suggested in the original thread but I wasn't sure how to use it in Sphinx. The css adjustments I made also affect all other bindings on the page but I think it actually looks better this way.

Once I know the final approach to take, I'll add it to the remaining pages.

joelostblom commented:
Wow it is indeed exciting to have the sliders working in the Altair docs! It might take me some time to review this PR properly, but my initial comments on the approach is that it makes sense. Did you copy/paste the vega-lite example or download it programmatically? It would be great if we would only need to keep track of the filenames for the examples in the vegalite docs/repo and pull down the latest version of them whenever the docs are built.

I agree with you that the new CSS looks better but is messes with how things look in the interactive Altair examples by centering the widgets and styling them in a way that will look different from what people would see when they run Altair on their own machines. Is is possible to apply this style tweak to only marks section? I think we could target the individual pages at least, even if there is no single css tag for the entire section.

mattijn commented:
A few non-substantive comments:
I really like it how you made it possible to render specific `properties` using the `altair-object-table`.

I also like it how you render VL-specs directly to make the docs more interactive. 

I think it is fine to keep these VL-specs defined within the Altair repository. Maybe the VL repository will soon introduce new things that Altair cannot render yet or we like to emphasize something different.

To continue on the last bit, it would be better if these VL specifications are Altair derived. In that case we can add a hyperlink below with "view Altair example" (like the "view in editor" at the VL-repo). The location of these Altair specifications is open for discussion, maybe in the interactivity section on the example page? Or in the doc-page of the parameters? Or somewhere else?

binste commented:
To summarise, based on your feedback I see the following 3 approaches we could take to include the interactive sliders:

1. Download the vega-lite specs from [this folder](https://github.com/vega/vega-lite/tree/v5.2.0/examples/specs) from their repo, put it into an HTML file, and include it using the raw html directive <- This is what I've done for the arc page. I could automate this with a script which would pull the specifications and create the html files. The script could be added to the make file for the documentation so it's easy to do this step. To make sure that the pulled specifications match the vega-lite version used in Altair, the script could use the git tags in the vl repo, e.g. "[https://raw.githubusercontent.com/vega/vega-lite/v{alt.VEGALITE_VERSION}/examples/specs/arc_params.vl.json](https://raw.githubusercontent.com/vega/vega-lite/v%7Balt.VEGALITE_VERSION%7D/examples/specs/arc_params.vl.json)"
2. Similar idea to 1 but creating a new Sphinx directive which just takes the file name of the specification such as "arc_params.vl.json" and which then performs the above steps without saving the specification first as an html file but directly including it. This way the examples would always be in sync without the need to run a script to update them.
3. Rewrite the vega-lite specs as Altair code. Ideally, the current altair_plot directive would then be extended to allow for collapsable code?

I can see the advantages of having these slider plots as Altair code. However, translating the code will take quite some time and I'm not familiar with creating or adapting Sphinx directives. 

In general, I'd like to get to the release of version 5 as quickly as possible to make all the great new features and the improved documentation available for everyone. I'd therefore suggest to go with option 1 for now as it will be fairly quick for me to implement. This would allow me to switch to tackling the creation of individual pages for each encoding as well which would be great for the release of version 5. I could create an enhancement issue stating that these examples could be converted to Altair code. What do you think?

ChristopherDavisUCI commented:
@mattijn @joelostblom I haven't been following the docs updates so I don't know how much would be involved, but I'm happy to try to help with translating from the Vega-Lite code to Altair code if you think that would be feasible/worthwhile.

I just checked out the [Arc page](https://binste.github.io/altair-docs/user_guide/marks/arc.html) and it looks great @binste!

mattijn commented:
That would be great @ChristopherDavisUCI! If you can prepare the Altair code for these mark-doc examples, that would be nice. I tried a bit as well, but I did not succeed in the arc example. I'd some errors that I could not get rid off.

mattijn commented:
Looks great! The styling is a trick in the documentation. Where should we place these? In the interaction section of the example gallery? Or a special section with parameters?

joelostblom commented:
Thanks everyone for working on this! Thinking out loud a bit, one hesitation I have regarding translating these to altair instead of going with option 1, is that maintenance will be harder and we will manually have to check when there are updates to the Vega-Lite doc pages to update our examples. If we go with option 1 then updates will be pulled automatically so that we will be notified about them right away (but we can still choose whether to commit them right away or later if the Altair library needs to get updated first).

I do agree that there is some advantage to seeing the examples in Altair code, and there are not that many mark pages that have elaborate use of sliders, so it doesn't seem too complex to maintain this. Overall i would say I am fine either way and don't feel strongly about this, but thought I would just raise this minor concern

ChristopherDavisUCI commented:
Hi @joelostblom, I haven't wrapped my head around the docs so maybe this doesn't make sense, but I was picturing just providing like a "related example" rather than an Altair version that produces the exact Vega-Lite spec.

joelostblom commented:
My understanding from the conversation above is that option 3 will replace the VegaLite spec with an Altair version and use the latter to generate the sliders and charts on the mark pages.

mattijn commented:
I like these examples in the docs and it is inspiring. But then coming too realize you can't reproduce it in Altair is somehow a bummer. 

Another reason why I think it's good to have Altair derived VL-specifications is because Altair adopts a top-level strategy for parameters instead of nested parameters. 

I think these Altair specifications are a bit tedious and large to be included as well within these mark documentation. So for me its fine to refer to another page where the Altair specification is written in full. This jupyterbook drop-down  style is fine too. 
How many examples are this in total? 7 or 8? Not too many I think

joelostblom commented:
Good points @mattijn , I agree it is a bit of a bummer for more advanced users to not be able to see how the charts can be made in Altair and I didn't think about the top level vs nested difference. And yes, there aren't that many examples overall and the Arc is the most complicated so it sounds like it makes sense if you keep trying to translate them @ChristopherDavisUCI .

@binste Do you want to go ahead and add the example Chris posted here to the arc page and then Chris can follow up with a PR for the other pages as he completes them? That way you don't have to wait for him to complete this PR and move on to the other parts you mentioned you would like to work on? Or would you prefer to do it some other way? For the dropdown panel, you can use `sphinx-panels` which I think we will use elsewhere in the docs later on as well (and it's what they use in jupyter-book too, https://sphinx-panels.readthedocs.io/en/latest/index.html)

mattijn commented:
To add. Based on the comment of @binste. I meant, a combination of 3 and 1:

- Rewrite the vega-lite specs as Altair code and use this Altair code to generate the VL-spec (this VL-spec is likely not exactly the same as the source from the VL-repo) and put this into an HTML file, and include it using the raw html directive as @binste have done already.
- Add this Altair code also as an example in the example gallery, so we can refer to it in the mark documentation page.

joelostblom commented:
Maybe I am missing something here, but if we have the altair spec, why do we need the HTML translation part? Can't we use the `.. altair-plot::` directive as for the other charts and the output should still render and be interactive? And on the marks landing page, we could link to the heading that contains the mark properties? Sorry if I am misunderstanding.

joelostblom commented:
Actually [there is already a `hide-code` option](https://github.com/altair-viz/altair/blob/0dea1f787ead644e88d1505498812b20a421afec/altair/sphinxext/altairplot.py#L32-L40) that we could use with the `..altair-plot` directive. It looks like it currently just [hides the code without a toggle](https://altair-viz.github.io/user_guide/internals.html#altair-chart-to-vega-lite-spec), but we could modify that using something like https://sphinx-togglebutton.readthedocs.io/en/latest/

mattijn commented:
If we add these VL examples as Altair code in the comments, than @binste can include them in this PR.
The following is from the VL-area example page: https://vega.github.io/vega-lite/docs/area.html#properties
```python
import altair as alt

interpolate_select = alt.binding_select(
    options=[
        "basis",
        "cardinal",
        "catmull-rom",
        "linear",
        "monotone",
        "natural",
        "step",
        "step-after",
        "step-before",
    ],
    name="interpolate"
)
interpolate_var = alt.param(bind=interpolate_select, value="linear")

tension_slider = alt.binding_range(min=0, max=1, step=0.05, name="tension")
tension_var = alt.param(bind=tension_slider, value=0)


source = pd.DataFrame({"u": [1, 2, 3, 4, 5, 6], "v": [28, 55, 42, 34, 36, 38]})

alt.Chart(source).mark_area(
    interpolate=interpolate_var, tension=tension_var
).encode(
    x='u',
    y='v'
).add_params(interpolate_var, tension_var)
```
<img width="462" alt="image" src="https://user-images.githubusercontent.com/5186265/204103920-83b11feb-359e-4a63-9470-c6550769d1e3.png">

ChristopherDavisUCI commented:
I will work on the slider examples for `mark_line`, `mark_point` and `mark_rect` next and post the Altair code here.  (If I'm misunderstanding something, please let me know!)

mattijn commented:
Here the VL bar-example (https://vega.github.io/vega-lite/docs/bar.html#properties):

```python
import altair as alt

corner_slider = alt.binding_range(min=0, max=50, step=1)
corner_var = alt.param(bind=corner_slider, value=0, name="cornerRadius")

source = pd.DataFrame(
    {
        "a": ["A", "B", "C", "D", "E", "F", "G", "H", "I"],
        "b": [28, 55, 43, 91, 81, 53, 19, 87, 52],
    }
)

alt.Chart(source).mark_bar(cornerRadius=corner_var).encode(
    x=alt.X("a:N", axis=alt.Axis(labelAngle=0)), y="b:Q"
).add_params(corner_var)
```
<img width="238" alt="image" src="https://user-images.githubusercontent.com/5186265/204104435-f4b7b6de-4557-4683-bb40-3dee45fc85f4.png">

ChristopherDavisUCI commented:
(By the way, I haven't run any of these through the `black` formatter or anything.)

Here is the [Line Mark Properties](https://vega.github.io/vega-lite/docs/line.html#properties) example:

```
import altair as alt
import pandas as pd

interpolate_select = alt.binding_select(
    options=[
      "basis",
      "cardinal",
      "catmull-rom",
      "linear",
      "monotone",
      "natural",
      "step",
      "step-after",
      "step-before"
    ],
    name="interpolate",
)
interpolate_var = alt.param(bind=interpolate_select, value="linear")

tension_slider = alt.binding_range(min=0, max=1, step=0.05, name="tension")
tension_var = alt.param(bind=tension_slider, value=0)

strokeWidth_slider = alt.binding_range(min=0, max=10, step=0.5, name="strokeWidth")
strokeWidth_var = alt.param(bind=strokeWidth_slider, value=2)

strokeCap_select = alt.binding_select(
    options=["butt", "round", "square"],
    name="strokeCap",
)
strokeCap_var = alt.param(bind=strokeCap_select, value="butt")

strokeDash_select = alt.binding_select(
    options=[
        [1, 0], 
        [8, 8], 
        [8, 4], 
        [4, 4], 
        [4, 2], 
        [2, 1], 
        [1, 1]
    ],
    name="strokeDash",
)
strokeDash_var = alt.param(bind=strokeDash_select, value=[1,0])

source = pd.DataFrame({"u": [1, 2, 3, 4, 5, 6], "v": [28, 55, 42, 34, 36, 38]})

alt.Chart(source).mark_line(
    interpolate=interpolate_var,
    tension=tension_var,
    strokeWidth=strokeWidth_var,
    strokeCap=strokeCap_var,
    strokeDash=strokeDash_var
).encode(
    x="u", y="v"
).add_params(
    interpolate_var, 
    tension_var, 
    strokeWidth_var,
    strokeCap_var,
    strokeDash_var
)
```
(As far as I can tell, the `tension` part does nothing here.  Not sure if it should be removed or not.)

<img width="452" alt="mark_line" src="https://user-images.githubusercontent.com/82750240/204104858-13ef35de-36ca-4df5-ac2b-93e93b50f033.png">

ChristopherDavisUCI commented:
Here is the [Mark Point Properties](https://vega.github.io/vega-lite/docs/point.html#properties) example:

```
import altair as alt
from vega_datasets import data

source = data.cars()

shape_select = alt.binding_select(
    options=[
      "arrow",
      "circle",
      "square",
      "cross",
      "diamond",
      "triangle",
      "triangle-up",
      "triangle-down",
      "triangle-right",
      "triangle-left",
      "wedge",
      "stroke",
      "M-1,-1H1V1H-1Z",
      "M0,.5L.6,.8L.5,.1L1,-.3L.3,-.4L0,-1L-.3,-.4L-1,-.3L-.5,.1L-.6,.8L0,.5Z"
    ],
    name="shape",
)
shape_var = alt.param(bind=shape_select, value="circle")

angle_slider = alt.binding_range(min=-360, max=360, step=1, name="angle")
angle_var = alt.param(bind=angle_slider, value=0)

size_slider = alt.binding_range(min=0, max=500, step=10, name="size")
size_var = alt.param(bind=size_slider, value=50)

strokeWidth_slider = alt.binding_range(min=0, max=10, step=0.5, name="strokeWidth")
strokeWidth_var = alt.param(bind=strokeWidth_slider, value=2)

alt.Chart(source).mark_point(
    shape=shape_var,
    angle=angle_var,
    size=size_var,
    strokeWidth=strokeWidth_var,
).encode(
    x="Horsepower:Q", 
    y="Miles_per_Gallon:Q"
).add_params(
    shape_var, 
    angle_var, 
    size_var, 
    strokeWidth_var
)
```

<img width="472" alt="mark_point" src="https://user-images.githubusercontent.com/82750240/204105718-5176c27a-afc4-469c-806a-a6bde0a00787.png">

mattijn commented:
And the [rule](https://vega.github.io/vega-lite/docs/rule.html#properties) example:
```python
import altair as alt

x_slider = alt.binding_range(min=1, max=100, step=1)
x_var = alt.param(bind=x_slider, value=35, name="x")

x2_slider = alt.binding_range(min=1, max=100, step=1)
x2_var = alt.param(bind=x2_slider, value=75, name="x2")

y_slider = alt.binding_range(min=1, max=100, step=1)
y_var = alt.param(bind=y_slider, value=25, name="y")

y2_slider = alt.binding_range(min=1, max=100, step=1)
y2_var = alt.param(bind=y2_slider, value=75, name="y2")

strokeWidth_slider = alt.binding_range(min=0, max=10, step=0.5)
strokeWidth_var = alt.param(
    bind=strokeWidth_slider, value=2, name="strokeWidth"
)

strokeCap_select = alt.binding_select(options=["butt", "round", "square"])
strokeCap_var = alt.param(
    bind=strokeCap_select, value="butt", name="strokeCap"
)

strokeDash_select = alt.binding_select(
    options=[[1, 0], [8, 8], [8, 4], [4, 4], [4, 2], [2, 1], [1, 1]]
)
strokeDash_var = alt.param(
    bind=strokeDash_select, value=[1, 0], name="strokeDash"
)

alt.Chart().mark_rule(
    color="orange",
    strokeWidth=strwidth_var,
    strokeCap=strkcp_var,
    strokeDash=strkdsh_var,
).encode(
    x=alt.datum(x_var, type="quantitative", scale=alt.Scale(domain=[0, 100])),
    y=alt.datum(y_var, type="quantitative", scale=alt.Scale(domain=[0, 100])),
    x2=alt.datum(x2_var),
    y2=alt.datum(y2_var),
).add_params(
    x_var,
    x2_var,
    y_var,
    y2_var,
    strokeWidth_var,
    strokeCap_var,
    strokeDash_var,
)
```
<img width="441" alt="image" src="https://user-images.githubusercontent.com/5186265/204106035-1d436320-38f9-4630-9902-2d994613b278.png">

ChristopherDavisUCI commented:
Here is the [mark rectangle properties](https://vega.github.io/vega-lite/docs/rect.html#properties) example:

```
import altair as alt

x_slider = alt.binding_range(min=1, max=100, step=1, name="x")
x_var = alt.param(bind=x_slider, value=25)

x2_slider = alt.binding_range(min=1, max=100, step=1, name="x2")
x2_var = alt.param(bind=x2_slider, value=75)

y_slider = alt.binding_range(min=1, max=100, step=1, name="y")
y_var = alt.param(bind=y_slider, value=25)

y2_slider = alt.binding_range(min=1, max=100, step=1, name="y2")
y2_var = alt.param(bind=y2_slider, value=75)

cornerRadius_slider = alt.binding_range(min=0, max=50, step=1)
cornerRadius_var = alt.param(bind=cornerRadius_slider, value=0, name="cornerRadius")

c = alt.Chart().mark_rect(
    cornerRadius=cornerRadius_var,
    color="orange"
).encode(
    x=alt.XDatum(x_var, type="quantitative", scale=alt.Scale(domain=[0,100])),
    x2=alt.X2Datum(x2_var),
    y=alt.XDatum(y_var, type="quantitative", scale=alt.Scale(domain=[0,100])),
    y2=alt.X2Datum(y2_var),
).add_params(
    x_var,
    x2_var,
    y_var,
    y2_var,
    cornerRadius_var
)
```

<img width="440" alt="mark_rect" src="https://user-images.githubusercontent.com/82750240/204106466-56cc5c1b-89a9-49f1-9c9d-ef295091f5d6.png">

@mattijn Does this syntax look right to you?  I had naively thought I might be able to simply do `x=x_var`, but I couldn't find anything simpler than what's above.

mattijn commented:
And the [text](https://vega.github.io/vega-lite/docs/text.html#properties) example:
```python
import numpy as np
import pandas as pd
import altair as alt

angle_slider = alt.binding_range(min=-180, max=180, step=1)
angle_var = alt.param(bind=angle_slider, value=0, name="angle")

dx_slider = alt.binding_range(min=-20, max=20, step=1)
dx_var = alt.param(bind=dx_slider, value=5, name="dx")

dy_slider = alt.binding_range(min=-20, max=20, step=1)
dy_var = alt.param(bind=dy_slider, value=0, name="dy")

xOffset_slider = alt.binding_range(min=-20, max=20, step=1)
xOffset_var = alt.param(bind=xOffset_slider, value=0, name="xOffset")

yOffset_slider = alt.binding_range(min=-20, max=20, step=1)
yOffset_var = alt.param(bind=yOffset_slider, value=0, name="yOffset")

fontSize_slider = alt.binding_range(min=1, max=36, step=1)
fontSize_var = alt.param(bind=fontSize_slider, value=14, name="fontSize")

limit_slider = alt.binding_range(min=0, max=150, step=1)
limit_var = alt.param(bind=limit_slider, value=0, name="limit")

align_select = alt.binding_select(options=["left", "center", "right"])
align_var = alt.param(bind=align_select, value="left", name="align")

baseline_select = alt.binding_select(
    options=["alphabetic", "top", "middle", "bottom"]
)
baseline_var = alt.param(bind=baseline_select, value="midle", name="baseline")

font_select = alt.binding_select(options=["sans-serif", "serif", "monospace"])
font_var = alt.param(bind=font_select, value="sans-serif", name="font")

fontWeight_select = alt.binding_select(options=["normal", "bold"])
fontWeight_var = alt.param(
    bind=fontWeight_select, value="normal", name="fontWeight"
)

fontStyle_select = alt.binding_select(options=["normal", "italic"])
fontStyle_var = alt.param(
    bind=fontStyle_select, value="normal", name="fontStyle"
)

source = pd.DataFrame(
    {
        "a": [30, 25, 70],
        "b": [28, 65, 43],
        "label": ["Andy", "Brian", "Charlie"],
    }
)


base = alt.Chart(source).encode(
    x=alt.X(
        "a:Q", axis=alt.Axis(labelAngle=0), scale=alt.Scale(domain=[0, 100])
    ),
    y=alt.Y("b:Q", scale=alt.Scale(domain=[0, 100])),
)

pts = base.mark_point()

text = base.mark_text(
    dx=dx_var,
    dy=dy_var,
    xOffset=xOffset_var,
    yOffset=yOffset_var,
    angle=angle_var,
    align=align_var,
    baseline=baseline_var,
    font=font_var,
    fontSize=fontSize_var,
    fontStyle=fontStyle_var,
    fontWeight=fontWeight_var,
    limit=limit_var,
).encode(text="label:N")

(pts + text).add_params(
    dx_var,
    dy_var,
    xOffset_var,
    yOffset_var,
    angle_var,
    align_var,
    baseline_var,
    font_var,
    fontSize_var,
    fontStyle_var,
    fontWeight_var,
    limit_var,
)
```
<img width="456" alt="image" src="https://user-images.githubusercontent.com/5186265/204106698-dca77a8e-e544-4dc7-815f-1eabee11a255.png">

ChristopherDavisUCI commented:
Thanks @mattijn!

I think this code will all look nicer once we have the attribute-based syntax from https://github.com/altair-viz/altair/pull/2592

mattijn commented:
I think so too @ChristopherDavisUCI!  
By the way, it did feel very tedious to write all these `alt.param()` with these `bind` variables and all these `.add_params()` after including them as expressions. Still room for improvement there as well.

ChristopherDavisUCI commented:
I felt the same way about this `param` and `bind` workflow.  I will take a look at that.

mattijn commented:
> @mattijn Does this syntax look right to you? I had naively thought I might be able to simply do x=x_var, but I couldn't find anything simpler than what's above.

Instead of `alt.X2Datum(x_var, ...)` you can do as well `alt.datum(x_var, ...)`, not sure if that is any better though.

mattijn commented:
> Actually [there is already a `hide-code` option](https://github.com/altair-viz/altair/blob/0dea1f787ead644e88d1505498812b20a421afec/altair/sphinxext/altairplot.py#L32-L40) that we could use with the `..altair-plot` directive. It looks like it currently just [hides the code without a toggle](https://altair-viz.github.io/user_guide/internals.html#altair-chart-to-vega-lite-spec), but we could modify that using something like https://sphinx-togglebutton.readthedocs.io/en/latest/

If it is possible to hide the code as well with the `..altair-plot`  directive, it's good too. These examples are very lengthy and this is what I was worried about when including the Altair code in the documentation. But somehow you also want to see the code to learn from or try it yourself. I'm fine with any approach, within the documentation using a toggle, as reference or something else, all fine, as long as the default is not to show the altair code directly. 
You are very good with this sphinx expression, so I trust you here😊

joelostblom commented:
Awesome, thanks for translating all those examples so quickly @mattijn and @ChristopherDavisUCI ! I looked into adding the toggle for the code and it took some time because the docutils documentation is (ironically) among the worst I have seen. But I finally figured it out and put it up in PR https://github.com/altair-viz/altair/pull/2725.

@binste You can go ahead and add the altair code directly in the documentation rst pages and use the `:hide-code:` option for rendering the chunks like this (you don't need to use `:links:`):

https://github.com/altair-viz/altair/blob/0dea1f787ead644e88d1505498812b20a421afec/doc/user_guide/times_and_dates.rst?plain=1#L51-L61  

Once the PR is merged, they will become toggleable as on this page https://joelostblom.github.io/altair-docs/user_guide/times_and_dates.html

binste commented:
Thank you all for your great help! It's really cool to see the Altair code (toggleable and hidden by default) right next to all the interactive examples :) I added all of them to this PR and updated the hosted documentation at https://binste.github.io/altair-docs/user_guide/marks/arc.html for you to review.

As suggested, I changed the css styling of the bindings to only apply to these examples and not the rest of the documentation by modifying the altairplot directive to support custom css class for the chart div.

Let me know if there are any other changes you think are necessary before merging this.

binste commented:
I also just sneeked in a small fix on the customizing visualizations page. I didn't open a separate PR as I needed the new altairplot functionality to add custom css classes introduced in this PR, hope that's ok!

Due to the new documentation layout (which is amazing!), this chart no longer showed up

![image](https://user-images.githubusercontent.com/23366411/205025923-206df2d0-8314-4d38-aacf-88f71b8dfbbd.png)

After the fix:

![image](https://user-images.githubusercontent.com/23366411/205025986-49db4e3e-528e-469f-9e10-a9e0e94d7363.png)

mattijn commented:
Looks great @binste!

mattijn reviewed (approved):
Minor comment: Maybe need to set a `x-overflow` for here? https://binste.github.io/altair-docs/user_guide/marks/point.html#point-mark-properties

![image](https://user-images.githubusercontent.com/5186265/205031662-d1b6f916-9b1d-46d7-b639-0213e99ac720.jpeg)

binste commented:
Yeah that one doesn't look great. However, not sure how to restrict it best as there is no obvious maximum width for the form containing those bindings. I played around with setting one for the whole form with

```css
.properties-example .vega-bindings {
  max-width: 400px
}
```
and then setting `overflow-x: hidden;` on the `vega-bind` div for that dropdown but it collapsed to 2 rows instead of clipping as expected.

joelostblom commented:
In the vega-lite docs they just use `max-width: 180px` on the `select` element (without anything for the overflow) and that seems to work here too for me 

![image](https://user-images.githubusercontent.com/4560057/205106737-cafcc2fc-6cc6-4891-befd-90f5d14e250b.png)

binste commented:
Simpler then expected... Thanks for the guidance! I added the css settings. Docs are updated so you can review.

joelostblom commented on doc/user_guide/marks/line.rst:
```diff
@@ -222,12 +286,9 @@ The ``interpolate`` property of a mark definition can be used to change line int
-    )
+    alt.Chart(source).mark_line(interpolate="monotone").encode(
+        x="date", y="price"
+    ).transform_filter(alt.datum.symbol == "GOOG")
```
Revert to the old syntax

joelostblom commented on doc/user_guide/marks/line.rst:
```diff
@@ -237,14 +298,12 @@ We can also set ``interpolate`` to ``"step-after"`` to create a step-chart.
+    alt.Chart(source).mark_line(interpolate="step-after").encode(
+        x="date", y="price"
+    ).transform_filter(alt.datum.symbol == "GOOG")
+
```
Revert to the old syntax

joelostblom commented on doc/user_guide/marks/line.rst:
```diff
@@ -263,33 +322,25 @@ By mapping geographic coordinate data to ``longitude`` and ``latitude`` channels
-        fill="lightgray",
-        stroke="white"
-    ).properties(
-        width=750,
-        height=500
-    ).project("albersUsa")
-
-    line = alt.Chart(source).mark_line().encode(
-        latitude="latitude:Q",
-        longitude="longitude:Q",
-        order = 'order'
-    ).transform_lookup(
-        lookup = 'airport',
-        from_ = lookup_data
+    source = pd.DataFrame(
+        {
+            "airport": ["SEA", "SFO", "LAX", "LAS", "DFW", "DEN", "ORD", "JFK"],
+            "order": [1, 2, 3, 4, 5, 6, 7, 8],
+        }
     )
 
-    background + line
-
-Line Config
-^^^^^^^^^^^
+    background = (
+        alt.Chart(states)
+        .mark_geoshape(fill="lightgray", stroke="white")
+        .properties(width=750, height=500)
+        .project("albersUsa")
+    )
 
-The ``line`` property of the top-level ``config`` object sets the default properties for all line marks. If mark property encoding channels are specified for marks, these config values will be overridden.
+    line = (
+        alt.Chart(source)
+        .mark_line()
+        .encode(latitude="latitude:Q", longitude="longitude:Q", order="order")
+        .transform_lookup(lookup="airport", from_=lookup_data)
+    )
+    background + line
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/parameter_charts/arc_params.html:
```diff
@@ -0,0 +1,89 @@
+<div id="arc-params">
```
We can remove this file and folder now since we have the altair specs instead

joelostblom commented on doc/user_guide/marks/point.rst:
```diff
@@ -4,110 +4,152 @@
-        x = 'IMDB_Rating:Q'
-    ) 
+    source = data.movies()
+    alt.Chart(source).mark_point().encode(x="IMDB_Rating:Q")
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/point.rst:
```diff
@@ -4,110 +4,152 @@
-        x = 'Horsepower:Q',
-        y = 'Miles_per_Gallon:Q'
-    )
+    alt.Chart(source).mark_point().encode(x="Horsepower:Q", y="Miles_per_Gallon:Q")
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/point.rst:
```diff
@@ -4,110 +4,152 @@
-    alt.Chart(source).mark_point(filled = True).encode(
-        x = 'Horsepower:Q',
-        y = 'Miles_per_Gallon:Q'
+    alt.Chart(source).mark_point(filled=True).encode(
+        x="Horsepower:Q", y="Miles_per_Gallon:Q"
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/point.rst:
```diff
@@ -4,110 +4,152 @@
-        x = 'Horsepower:Q',
-        y = 'Miles_per_Gallon:Q',
-        size = 'Acceleration:Q'
+        x="Horsepower:Q", y="Miles_per_Gallon:Q", size="Acceleration:Q"
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/point.rst:
```diff
@@ -4,110 +4,152 @@
-    ).properties(
-        height = alt.Step(50)
-    )
+        x="Horsepower:Q", y="Cylinders:O", yOffset="random:Q"
+    ).transform_calculate(random="random()").properties(height=alt.Step(50))
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/point.rst:
```diff
@@ -134,31 +176,26 @@ By mapping geographic coordinate data to ``longitude`` and ``latitude`` channels
     from vega_datasets import data
 
     airports = data.airports()
-    states = alt.topo_feature(data.us_10m.url, feature='states')
+    states = alt.topo_feature(data.us_10m.url, feature="states")
 
     # US states background
-    background = alt.Chart(states).mark_geoshape(
-        fill='lightgray',
-        stroke='white'
-    ).properties(
-        width=500,
-        height=300
-    ).project('albersUsa')
+    background = (
+        alt.Chart(states)
+        .mark_geoshape(fill="lightgray", stroke="white")
+        .properties(width=500, height=300)
+        .project("albersUsa")
+    )
 
     # airport positions on background
-    points = alt.Chart(airports).mark_circle(
-        size=10,
-        color='steelblue'
-    ).encode(
-        longitude='longitude:Q',
-        latitude='latitude:Q',
-        tooltip=['name', 'city', 'state']
+    points = (
+        alt.Chart(airports)
+        .mark_circle(size=10, color="steelblue")
+        .encode(
+            longitude="longitude:Q",
+            latitude="latitude:Q",
+            tooltip=["name", "city", "state"],
+        )
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/rect.rst:
```diff
@@ -36,23 +74,17 @@ For example, we can use ``rect`` to create an annotation ``layer`` that provides
     import altair as alt
     from vega_datasets import data
 
-    source =  data.cars()
+    source = data.cars()
 
-    point = alt.Chart(source).mark_point().encode(
-        x = 'Horsepower:Q',
-        y = 'Miles_per_Gallon:Q'
-    )
+    point = alt.Chart(source).mark_point().encode(x="Horsepower:Q", y="Miles_per_Gallon:Q")
 
-    rect = alt.Chart(source).mark_rect().encode(
-        y = 'max(Miles_per_Gallon)',
-        y2 = 'min(Miles_per_Gallon)',
-        opacity = alt.value(0.2)
+    rect = (
+        alt.Chart(source)
+        .mark_rect()
+        .encode(
+            y="max(Miles_per_Gallon)", y2="min(Miles_per_Gallon)", opacity=alt.value(0.2)
+        )
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/rule.rst:
```diff
@@ -21,11 +76,10 @@ We can use rules to show the average price of different stocks akin to ``tick``
-        y = 'mean(price):Q',
-        size = alt.value(2),
-        color = 'symbol:N'
+        y="mean(price):Q", size=alt.value(2), color="symbol:N"
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/rule.rst:
```diff
@@ -35,18 +89,8 @@ The fact that rule marks span the width or the height of a single view make them
     source = data.stocks()
 
     base = alt.Chart(source).properties(width=550)
-
-    line = base.mark_line().encode(
-        x='date',
-        y='price',
-        color='symbol'
-    )
-
-    rule = base.mark_rule().encode(
-        y='average(price)',
-        color='symbol',
-        size=alt.value(2)
-    )
+    line = base.mark_line().encode(x="date", y="price", color="symbol")
+    rule = base.mark_rule().encode(y="average(price)", color="symbol", size=alt.value(2))
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/rule.rst:
```diff
@@ -59,24 +103,17 @@ We can also use a rule mark to show global mean value over a histogram.
     source = data.movies.url
 
     base = alt.Chart(source)
-
-    bar = base.mark_bar().encode(
-        x=alt.X('IMDB_Rating:Q', bin=True, axis=None),
-        y='count()'
-    )
-
-    rule = base.mark_rule(color='red').encode(
-        x='mean(IMDB_Rating):Q',
-        size=alt.value(5)
-    )
+    bar = base.mark_bar().encode(x=alt.X("IMDB_Rating:Q", bin=True, axis=None), y="count()")
+    rule = base.mark_rule(color="red").encode(x="mean(IMDB_Rating):Q", size=alt.value(5))
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/rule.rst:
```diff
@@ -85,13 +122,5 @@ For example, we can use ``y`` and ``y2 ``show the ``"min"`` and ``"max"`` values
     source = data.cars()
 
     alt.Chart(source).mark_rule().encode(
-        x = 'Origin',
-        y = 'min(Horsepower)',
-        y2 = 'max(Horsepower)'
+        x="Origin", y="min(Horsepower)", y2="max(Horsepower)"
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/square.rst:
```diff
@@ -4,23 +4,22 @@
 
-The square config can contain any point mark properties (except ``type``, ``style``, and ``clip``).
\ No newline at end of file
+    alt.Chart(source).mark_square().encode(x="Horsepower:Q", y="Miles_per_Gallon:Q")
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/text.rst:
```diff
@@ -16,28 +113,28 @@ Text Table Heatmap
 
     source = data.cars()
 
-    base = alt.Chart(source).transform_aggregate(
-        num_cars='count()',
-        groupby=['Origin', 'Cylinders']
-    ).encode(
-        alt.X('Cylinders:O', scale=alt.Scale(paddingInner=0)),
-        alt.Y('Origin:O', scale=alt.Scale(paddingInner=0)),
+    base = (
+        alt.Chart(source)
+        .transform_aggregate(num_cars="count()", groupby=["Origin", "Cylinders"])
+        .encode(
+            alt.X("Cylinders:O", scale=alt.Scale(paddingInner=0)),
+            alt.Y("Origin:O", scale=alt.Scale(paddingInner=0)),
+        )
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/text.rst:
```diff
@@ -16,28 +113,28 @@ Text Table Heatmap
 
     source = data.cars()
 
-    base = alt.Chart(source).transform_aggregate(
-        num_cars='count()',
-        groupby=['Origin', 'Cylinders']
-    ).encode(
-        alt.X('Cylinders:O', scale=alt.Scale(paddingInner=0)),
-        alt.Y('Origin:O', scale=alt.Scale(paddingInner=0)),
+    base = (
+        alt.Chart(source)
+        .transform_aggregate(num_cars="count()", groupby=["Origin", "Cylinders"])
+        .encode(
+            alt.X("Cylinders:O", scale=alt.Scale(paddingInner=0)),
+            alt.Y("Origin:O", scale=alt.Scale(paddingInner=0)),
+        )
     )
 
     heatmap = base.mark_rect().encode(
-        color=alt.Color('num_cars:Q',
-            scale=alt.Scale(scheme='viridis'),
-            legend=alt.Legend(direction='horizontal')
+        color=alt.Color(
+            "num_cars:Q",
+            scale=alt.Scale(scheme="viridis"),
+            legend=alt.Legend(direction="horizontal"),
```
This is a good change and should be kept. This alignment of the parameters is more proper than the old version

joelostblom commented on doc/user_guide/marks/text.rst:
```diff
@@ -16,28 +113,28 @@ Text Table Heatmap
-            alt.value('black'),
-            alt.value('white')
-        )
+            alt.datum.num_cars > 100, alt.value("black"), alt.value("white")
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/text.rst:
```diff
@@ -50,45 +147,33 @@ You can also use ``text`` marks as labels for other marks and set offset (``dx``
     import altair as alt
     import pandas as pd
 
-    source = pd.DataFrame({ 
-        'a' : ['A', 'B', 'C'],
-        'b' : [28, 55, 43]
-    })
-
-    bar = alt.Chart(source).mark_bar().encode(
-        y = 'a:N',
-        x = alt.X('b:Q', scale = alt.Scale(domain = [0,60]))
-    )
+    source = pd.DataFrame({"a": ["A", "B", "C"], "b": [28, 55, 43]})
 
-    text = bar.mark_text( 
-                align = 'left', 
-                baseline = 'middle', 
-                dx = 3
-    ).encode(
-        text = 'b'
+    bar = (
+        alt.Chart(source)
+        .mark_bar()
+        .encode(y="a:N", x=alt.X("b:Q", scale=alt.Scale(domain=[0, 60])))
     )
+    text = bar.mark_text(align="left", baseline="middle", dx=3).encode(text="b")
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/text.rst:
```diff
@@ -50,45 +147,33 @@ You can also use ``text`` marks as labels for other marks and set offset (``dx``
-        x = 'Horsepower:Q',
-        y = 'Miles_per_Gallon:Q',
-        color = 'Origin:N',
-        text = 'Origin[0]:N'
+        x="Horsepower:Q", y="Miles_per_Gallon:Q", color="Origin:N", text="Origin[0]:N"
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/text.rst:
```diff
@@ -100,28 +185,23 @@ By mapping geographic coordinate data to ``longitude`` and ``latitude`` channels
 
     source = data.us_state_capitals()
 
-    background = alt.Chart(states).mark_geoshape(
-        fill="lightgray",
-        stroke="white"
-    ).properties(
-        width=750,
-        height=500
-    ).project("albersUsa")
-
-    line = alt.Chart(source).mark_text(dy = -10).encode(
-        latitude="lat:Q",
-        longitude="lon:Q",
-        text = 'city:N'
+    background = (
+        alt.Chart(states)
+        .mark_geoshape(fill="lightgray", stroke="white")
+        .properties(width=750, height=500)
+        .project("albersUsa")
     )
 
-    point = alt.Chart(source).mark_circle().encode(
-        latitude="lat:Q",
-        longitude="lon:Q",
-        color = alt.value('orange')
+    line = (
+        alt.Chart(source)
+        .mark_text(dy=-10)
+        .encode(latitude="lat:Q", longitude="lon:Q", text="city:N")
     )
 
-    background + line + point
+    point = (
+        alt.Chart(source)
+        .mark_circle()
+        .encode(latitude="lat:Q", longitude="lon:Q", color=alt.value("orange"))
+    )
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/tick.rst:
```diff
@@ -18,29 +26,20 @@ The following dot plot uses tick marks to show the distribution of precipitation
-Tick Config 
-^^^^^^^^^^^
-The ``tick`` property of the top-level ``config`` object sets the default properties for all tick marks. If mark property encoding channels are specified for marks, these config values will be overridden.
+    alt.Chart(source).mark_tick().encode(x="Horsepower:Q", y="Cylinders:O")
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/tick.rst:
```diff
@@ -50,9 +49,6 @@ Customizing Tick’s Size and Thickness
\ No newline at end of file
+    alt.Chart(source).mark_tick().encode(x="precipitation:Q").configure_tick(
+        thickness=2, bandSize=10
+    )
```
Revert to the old formatting

joelostblom commented on doc/user_guide/marks/trail.rst:
```diff
@@ -37,21 +45,22 @@ A Comet Chart showing changes between two states
         .transform_calculate(calculate="datum['1932'] - datum['1931']", as_="delta")
         .mark_trail()
         .encode(
-            x=alt.X('year:O', title=None),
-            y=alt.Y('variety:N', title='Variety'),
-            size=alt.Size('yield:Q', scale=alt.Scale(range=[0, 12]), legend=alt.Legend(values=[20, 60], title='Barley Yield (bushels/acre)')),
-            color=alt.Color('delta:Q', scale=alt.Scale(domainMid=0), legend=alt.Legend(title='Yield Delta (%)')),
-            tooltip=alt.Tooltip(['year:O', 'yield:Q']),
-            column=alt.Column('site:N', title='Site')
-
+            x=alt.X("year:O", title=None),
+            y=alt.Y("variety:N", title="Variety"),
+            size=alt.Size(
+                "yield:Q",
+                scale=alt.Scale(range=[0, 12]),
+                legend=alt.Legend(values=[20, 60], title="Barley Yield (bushels/acre)"),
+            ),
+            color=alt.Color(
+                "delta:Q",
+                scale=alt.Scale(domainMid=0),
+                legend=alt.Legend(title="Yield Delta (%)"),
+            ),
+            tooltip=alt.Tooltip(["year:O", "yield:Q"]),
+            column=alt.Column("site:N", title="Site"),
```
This looks better than the old version

joelostblom commented on doc/user_guide/marks/trail.rst:
```diff
@@ -37,21 +45,22 @@ A Comet Chart showing changes between two states
         .configure_view(stroke=None)
-        .configure_legend(orient='bottom', direction='horizontal')
-        .properties(title='Barley Yield comparison between 1932 and 1931')
+        .configure_legend(orient="bottom", direction="horizontal")
+        .properties(title="Barley Yield comparison between 1932 and 1931")
```
I would put the params on multiple lines for each of these

joelostblom reviewed (changes requested):
This update is so awesome! Thanks again @binste, your changes here will improve the docs notably and your contribution is really appreciated! I added some suggestions, there are many comments but most of them are about black messing up the formatting of the plot syntax.

mattijn commented on doc/user_guide/marks/geoshape.rst:
```diff
@@ -33,7 +33,7 @@ In the example above, Altair applies a default blue ``fill`` color and uses a de
 
```
There is one that I'm aware of: `filled`.

binste replied:
`filled` is already documented under the [standard mark properties](https://binste.github.io/altair-docs/user_guide/marks/index.html#color-properties) so I'll stick to Joel's original suggestion.

binste commented:
Happy to hear and thanks for the extensive feedback! :) I addressed all your comments. 

Indeed, I had formatted all the code examples with black to get a consistent style (always double quotes, no whitespaces around `=` for keyword arguments, ...) and I really like how it looks in general. But I can see the point that the code style which is mostly used across the documentation might be easier to read for the Altair syntax.

mattijn commented:
Looks great. I noticed the two following things:
- Can you have a look why this not renders: https://binste.github.io/altair-docs/user_guide/marks/square.html#scatter-plot-with-square. Seems same error as the fix in your last commit for the circle mark.
- All standard properties are shown for the line mark: https://binste.github.io/altair-docs/user_guide/marks/line.html#line-mark-properties, instead of the selection. Seems there should be an indent of 4 spaces for `:properties:`.

joelostblom reviewed (approved):
LGTM, merging! Thank again!
